packages/react-compose/src/types.ts

```diff
@@ -80,21 +95,29 @@ export type GenericDictionary = Record<string, any>;
  */
 export type ClassDictionary = Record<string, string>;
 
-export type ComposePreparedOptions<Props = {}> = {
+export type ComposePreparedOptions<TElementType extends React.ElementType = 'div', TProps = {}, TState = TProps> = {
   className: string;
-  classes: ClassDictionary;
+  classes: (undefined | ClassDictionary | ((state: GenericDictionary, slots: GenericDictionary) => ClassDictionary))[];
   displayName: string;
   displayNames: string[];
 
   mapPropsToStylesPropsChain: ((props: object) => object)[];
-  render: ComposeRenderFunction;
+  render: ComposeRenderFunction<TElementType, TProps>;
 
-  handledProps: (keyof Props)[];
+  handledProps: (keyof TProps)[];
   overrideStyles: boolean;
 
   slots: Record<string, React.ElementType> & { __self: React.ElementType };
-  mapPropsToSlotPropsChain: ((props: Props) => Record<string, object>)[];
-
+  mapPropsToSlotPropsChain: ((props: TProps) => Record<string, object>)[];
   resolveSlotProps: <P>(props: P) => Record<string, object>;
-  shorthandConfig: ShorthandConfig<Props>;
+
+  shorthandConfig: ShorthandConfig<TProps>;
+
+  resolve: (
```

Comment:
We should not have both `resolveSlotProps` and `resolveProps`. `compose` is already confusing enough